RecommenderSystems/dcn/dcn_train_eval.py

```diff
@@ -0,0 +1,695 @@
+        "--table_size_array",
+        type=int_list,
+        help="Embedding table size array for sparse fields",
+        required=True,
+    )
+    parser.add_argument(
+        "--persistent_path", type=str, required=True, help="path for persistent kv store"
+    )
+    parser.add_argument("--store_type", type=str, default="cached_host_mem")
+    parser.add_argument("--cache_memory_budget_mb", type=int, default=8192)
+
+    parser.add_argument("--amp", action="store_true", help="Run model with amp")
+    parser.add_argument("--loss_scale_policy", type=str, default="static", help="static or dynamic")
+
+    args = parser.parse_args()
+
+    if print_args and flow.env.get_rank() == 0:
+        _print_args(args)
+    return args
+
+
+def _print_args(args):
+    """Print arguments."""
+    print("------------------------ arguments ------------------------", flush=True)
+    str_list = []
+    for arg in vars(args):
+        dots = "." * (48 - len(arg))
+        str_list.append("  {} {} {}".format(arg, dots, getattr(args, arg)))
+    for arg in sorted(str_list, key=lambda x: x.lower()):
+        print(arg, flush=True)
+    print("-------------------- end of arguments ---------------------", flush=True)
+
+
+class DCNDataReader(object):
+    """A context manager that manages the creation and termination of a
+    :class:`petastorm.Reader`.
+    """
+
+    def __init__(
+        self,
+        parquet_file_url_list,
+        batch_size,
+        num_epochs=1,
+        shuffle_row_groups=True,
+        shard_seed=2022,
+        shard_count=1,
+        cur_shard=0,
+    ):
+        self.parquet_file_url_list = parquet_file_url_list
+        self.batch_size = batch_size
+        self.num_epochs = num_epochs
+        self.shuffle_row_groups = shuffle_row_groups
+        self.shard_seed = shard_seed
+        self.shard_count = shard_count
+        self.cur_shard = cur_shard
+
+        fields = ["Label"]
+        fields += [f"I{i+1}" for i in range(num_dense_fields)]
+        fields += [f"C{i+1}" for i in range(num_sparse_fields)]
+        self.fields = fields
+        self.num_fields = len(fields)
+
+    def __enter__(self):
+        self.reader = make_batch_reader(
+            self.parquet_file_url_list,
+            workers_count=2,
+            shuffle_row_groups=self.shuffle_row_groups,
+            num_epochs=self.num_epochs,
+            shard_seed=self.shard_seed,
+            shard_count=self.shard_count,
+            cur_shard=self.cur_shard,
+        )
+        self.loader = self.get_batches(self.reader)
+        return self.loader
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        self.reader.stop()
+        self.reader.join()
+
+    def get_batches(self, reader, batch_size=None):
+        if batch_size is None:
+            batch_size = self.batch_size
+
+        tail = None
+
+        for rg in reader:
+            rgdict = rg._asdict()
+            rglist = np.array([rgdict[field] for field in self.fields])
```

Comment:
np.array will force to convert dtype to float64.
more information: 
https://github.com/Oneflow-Inc/models/pull/339/files#r869078975